chore: Improve error context in controller error handling

moko-poi opened the pull request:
chore: improve error handling with contextual information in controller packages

Fixes #2630

**Description**

This PR improves error handling in controller packages by wrapping errors with contextual information. When errors occur deep in the call stack, it's now easier to understand which operation failed and with what parameters.

**Changes made:**
- Added error wrapping to 15 locations across 9 controller files
- Included resource identifiers (NodePool name, NodeClaim name, Node name) in error messages
- Followed existing error message patterns using `fmt.Errorf(..., %w, err)` format
- Used consistent verb forms (present participle: getting, validating, creating, etc.)

**Example improvements:**

Before:
```go
if err != nil {
    return "", err  // No context about which operation failed
}
```

After:
```go
if err != nil {
    return "", fmt.Errorf("getting instance types for nodepool %q, %w", nodePool.Name, err)
}
```

**Files Changed:**

| File                                                 | Lines Changed | Description                                             |
|------------------------------------------------------|---------------|---------------------------------------------------------|
| pkg/controllers/nodeclaim/disruption/drift.go        | 2             | Add context for instance type and drift checking errors |
| pkg/controllers/disruption/types.go                  | 2             | Add context for node and pod validation errors          |
| pkg/controllers/state/statenode.go                   | 2             | Add context for pod retrieval errors                    |
| pkg/controllers/disruption/helpers.go                | 1             | Add context for nodepool map building error             |
| pkg/controllers/disruption/queue.go                  | 1             | Add context for replacement nodeclaim creation error    |
| pkg/controllers/disruption/validation.go             | 3             | Add context for candidate validation errors             |
| pkg/controllers/disruption/multinodeconsolidation.go | 1             | Add context for instance type filtering error           |
| pkg/controllers/nodeclaim/lifecycle/liveness.go      | 1             | Add context for nodeclaim deletion error                |
| pkg/controllers/nodeclaim/lifecycle/registration.go  | 2             | Add context for nodepool get/patch errors               |

**How was this change tested?**

- All changes pass `make verify`
- `golangci-lint` reports 0 issues
- `go vet` passes without errors
- Error messages follow existing codebase patterns

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

moko-poi committed:
chore: add error context to controller error handling

moko-poi committed:
fix: update test expectations for error context wrapping

ryan-mist commented on pkg/controllers/disruption/helpers.go:
```diff
@@ -176,7 +176,7 @@ func GetCandidates(ctx context.Context, cluster *state.Cluster, kubeClient clien
 	nodePoolMap, nodePoolToInstanceTypesMap, err := BuildNodePoolMap(ctx, kubeClient, cloudProvider)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("building nodepool map, %w", err)
```
The only error case returned in `BuildNodePoolMap` is `fmt.Errorf("listing node pools, %w", err)` ([here](https://github.com/kubernetes-sigs/karpenter/blob/main/pkg/controllers/disruption/helpers.go#L198)). 

Adding a `"building nodepool map"` to the error message here might not add too much value since the core issue if this throws an error is already captured.

moko-poi replied:
1c83049
Thanks for the review! You're absolutely right - I've removed the redundant error wrapping in both locations. The underlying functions already provide sufficient context, so the additional wrapping wasn't adding value. Changes have been pushed.

moko-poi committed:
refactor: remove redundant error wrapping in disruption package

Remove duplicate error context wrapping where underlying functions
already provide sufficient error information. This addresses review
feedback to avoid excessive error message nesting.

moko-poi commented on pkg/controllers/disruption/queue.go:
```diff
@@ -287,7 +287,7 @@ func (q *Queue) markDisrupted(ctx context.Context, cmd *Command) ([]*Candidate,
 	nodeClaimNames, err := q.provisioner.CreateNodeClaims(ctx, lo.Map(cmd.Replacements, func(r *Replacement, _ int) *pscheduling.NodeClaim { return r.NodeClaim }), provisioning.WithReason(strings.ToLower(string(cmd.Reason()))))
 	if err != nil {
-		return err
+		return fmt.Errorf("creating replacement nodeclaims, %w", err)
```
1c83049
Thanks for the review! You're absolutely right - I've removed the redundant error wrapping in both locations. The underlying functions already provide sufficient context, so the additional wrapping wasn't adding value. Changes have been pushed.

ryan-mist commented on pkg/controllers/disruption/types.go:
```diff
@@ -98,7 +98,11 @@ func NewCandidate(ctx context.Context, kubeClient client.Client, recorder events
+		if node.Node != nil {
+			nodeName = node.Node.Name
+		}
+		return nil, fmt.Errorf("validating node %q for disruption, %w", nodeName, err)
```
Looks to be helpful, although we may want to change the errors in `ValidateNodeDisruptable` to have the node name instead of here. I don't have a preference either way but noting for other reviewers.

Errors can occur when the StateNode cannot be disrupted, currently we would see one of the error messages [here](https://github.com/kubernetes-sigs/karpenter/blob/release-v1.8.x/pkg/controllers/state/statenode.go#L202), where the node name isn't mentioned.

ryan-mist commented:
No worries about adding them, was just wondering if you had already had them around. On some of the new error messages I added some examples of new logs to help the next reviewer.

moko-poi commented on pkg/controllers/disruption/multinodeconsolidation.go:
```diff
@@ -218,7 +218,7 @@ func filterOutSameInstanceType(replacement *Replacement, consolidate []*Candidat
 	replacement.NodeClaim, err = replacement.RemoveInstanceTypeOptionsByPriceAndMinValues(replacement.Requirements, maxPrice)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("filtering instance types by price, %w", err)
```
Thanks for confirming! The error message will now show both the operation context ("filtering instance types by price") and the underlying reason (specific labels that failed minValues), which should make debugging easier.

moko-poi commented on pkg/controllers/nodeclaim/lifecycle/registration.go:
```diff
@@ -104,7 +104,7 @@ func (r *Registration) updateNodePoolRegistrationHealth(ctx context.Context, nod
 		nodePool := &v1.NodePool{}
 		if err := r.kubeClient.Get(ctx, types.NamespacedName{Name: nodePoolName}, nodePool); err != nil {
-			return err
+			return fmt.Errorf("getting nodepool %q, %w", nodePoolName, err)
```
Good question! I checked the Kubernetes client-go error messages:

- **NotFound errors** include the resource type and name: `NodePool.karpenter.sh "my-nodepool" not found`
- **Other errors** (timeouts, permissions, etc.) typically don't include the name

While there's some redundancy for NotFound cases, I think it's worth keeping the wrapper because:
1. It adds operation context ("getting nodepool") which the underlying error lacks
2. It ensures the nodepool name is always included, even for non-NotFound errors
3. It's consistent with other places in the codebase (e.g., `pkg/utils/pod/pod.go:31`)

The slight redundancy seems acceptable for better debuggability. What do you think?

moko-poi commented on pkg/controllers/state/statenode.go:
```diff
@@ -92,7 +92,7 @@ func (n StateNodes) Pods(ctx context.Context, kubeClient client.Client) ([]*core
 		p, err := node.Pods(ctx, kubeClient)
 		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf("getting pods for node %q, %w", node.Node.Name, err)
```
caba367
Great suggestion! I've already implemented exactly what you recommended in my earlier commit. The error context is now in the underlying `GetPods` function (`pkg/utils/node/node.go:108`):

```go
return nil, fmt.Errorf("listing pods for node %q, %w", node.Name, err)
```

And I removed the redundant wrapping in StateNodes.Pods() and ValidatePodsDisruptable() to avoid duplication. This way, all callers of [GetPods](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) (including the termination controller) automatically get the node name in error messages.

moko-poi committed:
refactor: move error context to lower layers per review feedback

moko-poi committed:
fix: add pricingController
